### PR TITLE
feat(forms-web-app): added full-planning local-planning-department page

### PR DESCRIPTION
**Pull request by swdpcomputing**
## Ticket Number

https://pins-ds.atlassian.net/browse/AS-3782

## Description of change
added full-planning local-planning-department page

## Checklist

- [ ] Requires infrastructure changes
- [ ] If adding or remove environment variables (e.g. in `docker-compose.yaml`) then I have updated the appropriate Helm chart
- [ ] I have updated the documentation accordingly
- [x] My commit history in this PR is linear
- [x] New features have tests
- [ ] Breaking change (team conversation required)

## Important

Please do not merge from `master` (please only [rebase](https://github.com/foundry4/appeal-planning-decision/wiki/An-intro-to-Git-Rebase)). This keeps the history linear and easier to debug.

### CODE REVIEW

**Review by nssdesikan — CHANGES_REQUESTED** [review]
@swdpcomputing This screen is common for Householder appeal and Full planning. So the the new changes need not go under full-planning folder.
I will review the PR further once this is resolved

File: packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    res.render(VIEW.BEFORE_YOU_START.LOCAL_PLANNING_DEPARTMENT, {
+      appeal,
+      departments: departmentsToNunjucksItems(departments, lpaName),
+      errors,
+      errorSummary: [{ text: e.toString(), href: '#' }],
+    });
+    return;
+  }
+
+  if (!errors['local-planning-department']) {
```
Comment (nssdesikan) [inline packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js:71]: This check is not required

File: packages/forms-web-app/src/views/before-you-start/local-planning-department.njk [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    {{ govukErrorSummary({
+      titleText: "There is a problem",
+      errorList: errorSummary
+    }) }}
+  {% endif %}
+
+  <div class="govuk-grid-row">
+    <div class="govuk-grid-column-two-thirds">
+
+      <p class="govuk-visually-hidden" data-cy="eligible-departments">{{ eligibleDepartments }}</p>
```
Comment (nssdesikan) [inline packages/forms-web-app/src/views/before-you-start/local-planning-department.njk:27]: I may be missing something. But I could not find the use of these hidden fields.

File: packages/forms-web-app/src/views/before-you-start/local-planning-department.njk [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+              text: errors['local-planning-department'].msg
+            },
+            items: departments
+          }) }}
+
+        </div>
+
+        {{ govukButton({
+          text: "Continue",
+          type: "Submit",
```
Comment (nssdesikan) [inline packages/forms-web-app/src/views/before-you-start/local-planning-department.njk:56]: Submit should be in all lowercase (accessibility check) 
```suggestion
          type: "submit",
```

File: packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+      };
+
+      await localPlanningDepartmentController.postPlanningDepartment(mockRequest, res);
+
+      expect(createOrUpdateAppeal).toHaveBeenCalledWith({
+        ...appeal,
+        lpaCode: 'lpaCode1',
+      });
+
+      // expect(res.redirect).toBeCalledWith(`/${VIEW.}`); // Future Planning Application Decision Page
```
Comment (nssdesikan) [inline packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js:124]: Its better not to have commented out code. We can implement this when we play the future card

File: packages/forms-web-app/__tests__/unit/controllers/full-planning/local-planning-department.test.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
+const localPlanningDepartmentController = require('../../../../src/controllers/full-planning/local-planning-department');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/full-planning/local-planning-department.test.js:1]: This import is required for the tests to work.

[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/full-planning/local-planning-department.test.js:1]: Conflict merge error. removed.

File: packages/forms-web-app/__tests__/unit/routes/full-planning/local-planning-department.test.js [hunk 6]
```diff
@@ -0,0 +1,35 @@
+const { get, post } = require('../router-mock');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/full-planning/local-planning-department.test.js:1]: Conflict merge error. removed.

File: packages/forms-web-app/__tests__/unit/validators/full-planning/local-planning-department.test.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
+jest.mock('../../../../src/services/department.service');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/full-planning/local-planning-department.test.js:1]: Conflict merge error. removed.

File: packages/forms-web-app/src/controllers/full-planning/local-planning-department.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
+const logger = require('../../lib/logger');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/full-planning/local-planning-department.js:1]: Conflict merge error. removed.

File: packages/forms-web-app/src/routes/full-planning/index.js [hunk 9]
```diff
@@ -0,0 +1,7 @@
+const express = require('express');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/routes/full-planning/index.js:1]: Conflict merge error. removed.

File: packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
+const logger = require('../../lib/logger');
+const { getDepartmentFromName } = require('../../services/department.service');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js:2]: No performance gain, but certainly better formatting. Edited.

File: packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js (hunk 1 above)
Comment (swdpcomputing) [reply packages/forms-web-app/src/controllers/before-you-start/local-planning-department.js:71]: Removed.

File: packages/forms-web-app/src/views/before-you-start/local-planning-department.njk (hunk 2 above)
Comment (swdpcomputing) [reply packages/forms-web-app/src/views/before-you-start/local-planning-department.njk:27]: Me neither. I was keeping to earlier versions of the page as the goal is to keep them similar, so I'm glad you've flagged this. Removed.

File: packages/forms-web-app/src/views/before-you-start/local-planning-department.njk (hunk 3 above)
Comment (swdpcomputing) [reply packages/forms-web-app/src/views/before-you-start/local-planning-department.njk:56]: Edited.

File: packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js (hunk 4 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js:124]: Edited.

File: packages/forms-web-app/src/routes/before-you-start/index.js [hunk 11]
```diff
@@ -0,0 +1,7 @@
+const express = require('express');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/routes/before-you-start/index.js:1]: Added. Strange that this wasn't kicking off the code coverage.

File: packages/forms-web-app/src/validators/before-you-start/local-planning-department.js [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+  }
+
+  return department;
+};
+
+const rules = () => {
+  return [
+    body('local-planning-department')
+      .notEmpty()
+      .withMessage('Which local planning department dealt with your planning application?')
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/validators/before-you-start/local-planning-department.js:22]: Edited.

File: packages/forms-web-app/src/validators/before-you-start/local-planning-department.js [hunk 13]
```diff
@@ -0,0 +1,30 @@
+const { body } = require('express-validator');
+const { getDepartmentData } = require('../../services/department.service');
+
+const validateDepartment = async (department) => {
+  const { departments, eligibleDepartments } = await getDepartmentData();
+
+  if (!departments.includes(department)) {
+    throw new Error('Which local planning department dealt with your planning application?');
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/src/validators/before-you-start/local-planning-department.js:8]: Edited.

File: packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+const { testExpressValidatorMiddleware } = require('../validation-middleware-helper');
+
+const departmentsData = {
+  departments: ['lpa1', 'lpa2'],
+  eligibleDepartments: ['lpa1'],
+};
+
+getDepartmentData.mockResolvedValue(departmentsData);
+
+describe('validators/planning-department', () => {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js:14]: Edited.

File: packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+    it('is configured with the expected rules', () => {
+      const rule = rules()[0].builder.build();
+
+      expect(rules().length).toEqual(1);
+      expect(rule.fields).toEqual(['local-planning-department']);
+      expect(rule.locations).toEqual(['body']);
+      expect(rule.optional).toBeFalsy();
+      expect(rule.stack).toHaveLength(3);
+      expect(rule.stack[0].message).toEqual(
+        'Which local planning department dealt with your planning application?'
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js:25]: Edited.

File: packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+      );
+    });
+  });
+  describe('validator', () => {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js:29]: No, general term of validate widely used in repo.

File: packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+      {
+        title: 'local planning department is not provided',
+        given: () => ({
+          body: {},
+        }),
+        expected: (result) => {
+          expect(result.errors).toHaveLength(1);
+          expect(result.errors[0].location).toEqual('body');
+          expect(result.errors[0].msg).toEqual(
+            'Which local planning department dealt with your planning application?'
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js:63]: Edited.

File: packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+        given: () => ({
+          body: {
+            'local-planning-department': 'lpa23',
+          },
+        }),
+        expected: (result) => {
+          expect(result.errors).toHaveLength(1);
+          expect(result.errors[0].location).toEqual('body');
+          expect(result.errors[0].msg).toEqual(
+            'Which local planning department dealt with your planning application?'
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/validators/before-you-start/local-planning-department.test.js:79]: Edited.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+const localPlanningDepartmentController = require('../../../../src/controllers/before-you-start/local-planning-department');
+const { validationErrorHandler } = require('../../../../src/validators/validation-error-handler');
+const fetchExistingAppealMiddleware = require('../../../../src/middleware/fetch-existing-appeal');
+const {
+  rules: localPlanningDepartmentValidationRules,
+} = require('../../../../src/validators/before-you-start/local-planning-department');
+
+jest.mock('../../../../src/validators/before-you-start/local-planning-department');
+
+describe('routes/eligibility/planning-department', () => {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js:11]: Edited.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+  beforeEach(() => {
+    // eslint-disable-next-line global-require
+    require('../../../../src/routes/before-you-start/local-planning-department');
+  });
+
+  afterEach(() => {
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js:17]: Removed.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js [hunk 21]
```diff
@@ -0,0 +1,18 @@
...
+
+describe('routes/before-you-start/index', () => {
+  beforeEach(() => {
+    jest.resetModules();
+
+    // eslint-disable-next-line global-require
+    require('../../../../src/routes/before-you-start');
+  });
+
+  afterEach(() => {
```
Comment (nssdesikan) [inline packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:11]: We don't need afterEach as there is only one test exists

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+    jest.resetAllMocks();
+  });
+
+  it('should define the expected routes', () => {
+    expect(use.mock.calls.length).toBe(1);
```
Comment (nssdesikan) [inline packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:16]: We may another expectation that checks for call to specific router
```suggestion
    expect(use.mock.calls.length).toBe(1);
    expect(use).toHaveBeenCalledWith('/', localPlanningDepartmentRouter);
```

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+const localPlanningDepartmentController = require('../../../../src/controllers/before-you-start/local-planning-department');
+const { validationErrorHandler } = require('../../../../src/validators/validation-error-handler');
+const fetchExistingAppealMiddleware = require('../../../../src/middleware/fetch-existing-appeal');
+const {
+  rules: localPlanningDepartmentValidationRules,
+} = require('../../../../src/validators/before-you-start/local-planning-department');
+
+jest.mock('../../../../src/validators/before-you-start/local-planning-department');
+
+describe('routes/eligibility/local-planning-department', () => {
```
Comment (nssdesikan) [inline packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js:11]: Incorrect path is referred here
```suggestion
describe('routes/before-you-start/local-planning-department', () => {
```

File: packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js (hunk 4 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js:124]: Edited.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js [hunk 24]
```diff
@@ -0,0 +1,18 @@
+const { use } = require('../router-mock');
+
+describe('routes/before-you-start/index', () => {
+  beforeEach(() => {
+    jest.resetModules();
```
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:5]: Removed.

[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:5]: Removed

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js (hunk 21 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:11]: Removed

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js (hunk 22 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:16]: Edited.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js (hunk 23 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/local-planning-department.test.js:11]: Edited.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js (hunk 22 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:16]: Leaving this test as it is to keep consistency with other sections.

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js (hunk 24 above)
[the remark this answers is not in the document]
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:5]: Leaving this test as it is to keep consistency with other sections.

File: packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js (hunk 4 above)
Comment (nssdesikan) [reply packages/forms-web-app/__tests__/unit/controllers/before-you-start/local-planning-department.test.js]: Commented out code still exists

File: packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js (hunk 21 above)
Comment (swdpcomputing) [reply packages/forms-web-app/__tests__/unit/routes/before-you-start/index.test.js:11]: Leaving this test as it is to keep consistency with other sections.

**Review by nssdesikan — APPROVED** [review]

**Pull request merged**